src/plugins/extra/smartquotes.rs

```diff
@@ -0,0 +1,517 @@
+    md.add_rule::<SmartQuotesRule<'‘', '’', '“', '”'>>();
+}
+
+/// Simplified Node type that only holds the info we need
+///
+/// To replace quotes, we'll be iterating forward and backward over the nodes in
+/// our document tree. The `Node` class doesn't provide a mechanism to do this
+/// efficiently, and in any case we only care about certain parts of the
+/// information. This struct will be used to build a flat view of the document;
+/// the `Irrelevant` variant serves as a "filler" so that the indexes of the
+/// entries line up correctly with the order we see during tree traversal.
+#[derive(Debug)]
+enum FlatToken<'a> {
+    LineBreak,
+    Text {
+        content: &'a str,
+        nesting_level: u32,
+    },
+    Irrelevant,
+}
+
+/// A simple enum to distinguish single and double quotes
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+enum QuoteType {
+    Single,
+    Double,
+}
+
+/// Holds information about quotes we have encountered thus far.
+///
+/// These quotes may or may not be used to close a pair further down the line.
+/// The different fields thus hold all the information we need to a) decide
+/// whether or not to match them up with anoter quote we encounter, and b) to
+/// perform the correct replacement, should be indeed use this quote to close a
+/// pair.
+#[derive(Debug)]
+struct QuoteMarker {
+    /// The iteration index of the node in which this quote was found.
+    ///
+    /// This is the index at which this quote's `Node` appears in a pre-order
+    /// depth-first walk of the document tree. Since we can only _modify_ nodes
+    /// during a walk, we rely on this index to tell us which nodes to modify.
+    walk_index: usize,
+    /// The position of the quote within node's `content`
+    quote_position: usize,
+    /// Whether this is a single or a double quote
+    quote_type: QuoteType,
+    /// Nesting level of the containing token
+    ///
+    /// This is the nesting of the containing `Node` within the document tree.
+    /// It is used to decide which quotes can be matched up.
+    level: u32,
+}
+
+/// Description of a single quote replacement to be executed
+///
+/// As described above, we have to compute the replacements in a first step that
+/// treats the entire document tree read-only. Only then can we perform the
+/// actual replacements. This `struct` holds the information we need to perform
+/// the replacement of a single quote character during a `walk_mut`.
+struct ReplacementOp {
+    walk_index: usize,
+    quote_position: usize,
+    quote: char,
+}
+
+impl<
+        const OPEN_SINGLE_QUOTE: char,
+        const CLOSE_SINGLE_QUOTE: char,
+        const OPEN_DOUBLE_QUOTE: char,
+        const CLOSE_DOUBLE_QUOTE: char,
+    > CoreRule
+    for SmartQuotesRule<
+        OPEN_SINGLE_QUOTE,
+        CLOSE_SINGLE_QUOTE,
+        OPEN_DOUBLE_QUOTE,
+        CLOSE_DOUBLE_QUOTE,
+    >
+{
+    fn run(root: &mut Node, _: &MarkdownIt) {
+        let text_tokens = all_text_tokens(root);
```

Comment:
JS has pretest that returns early if input has no quotes at all (to improve performance by skipping the heavy logic in most cases)

maybe `all_text_tokens` can also check if quotes (`'` or `"`) are found, and return None/empty vec as a shortcut

needs benchmarks though